api/grpcserver/post_client.go

```diff
@@ -0,0 +1,101 @@
+package grpcserver
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	pb "github.com/spacemeshos/api/release/go/spacemesh/v1"
+
+	"github.com/spacemeshos/go-spacemesh/common/types"
+)
+
+// postClient represents a connection to a PoST service.
+//
+// It uses the grpc interface of the node to send commands to the post service.
+// Additionally if instructed it will start the post service and connect it to
+// the node.
+type postClient struct {
+	service *PostService
+
+	closed chan struct{}
+}
+
+func newPostClient(service *PostService) *postClient {
+	return &postClient{
+		service: service,
+		closed:  make(chan struct{}),
+	}
+}
+
+func (pc *postClient) Proof(ctx context.Context, challenge []byte) (*types.Post, *types.PostMetadata, error) {
+	ch := make(chan *pb.ServiceResponse, 1)
+	cmd := postCommand{
+		req: &pb.NodeRequest{
+			Kind: &pb.NodeRequest_GenProof{
+				GenProof: &pb.GenProofRequest{
+					Challenge: challenge,
+				},
+			},
+		},
+		ch: ch,
+	}
+
+	select {
+	case <-pc.closed:
+		return nil, nil, fmt.Errorf("post client closed")
+	default:
+	}
+
+	if err := pc.service.Send(ctx, cmd); err != nil {
+		return nil, nil, fmt.Errorf("failed to send request: %w", err)
+	}
+
+	for {
+		select {
+		case <-pc.closed:
+			return nil, nil, fmt.Errorf("post client closed")
+		case <-ctx.Done():
+			return nil, nil, ctx.Err()
+		case resp := <-ch:
+			proofResp, ok := resp.GetKind().(*pb.ServiceResponse_GenProof)
+			if !ok {
+				return nil, nil, fmt.Errorf("unexpected response of type: %T", resp.GetKind())
+			}
+			switch proofResp.GenProof.GetStatus() {
+			case pb.GenProofStatus_GEN_PROOF_STATUS_OK:
+				if proofResp.GenProof.GetProof() == nil {
+					select {
+					case <-ctx.Done():
+						return nil, nil, ctx.Err()
+					case <-time.After(2 * time.Second): // TODO(mafa): make polling interval configurable
+					}
+				}
+
+				proof := proofResp.GenProof.GetProof()
+				meta := proofResp.GenProof.GetMetadata()
```

Comment:
There is a helper for this: `resp.GetGenProof()`:
```suggestion
			genProof := resp.GetGenProof()
			if genProof == nil {
				return nil, nil, fmt.Errorf("unexpected response of type: %T", resp.GetKind())
			}
			switch genProof.GetStatus() {
			case pb.GenProofStatus_GEN_PROOF_STATUS_OK:
				if genProof.GetProof() == nil {
					select {
					case <-ctx.Done():
						return nil, nil, ctx.Err()
					case <-time.After(2 * time.Second): // TODO(mafa): make polling interval configurable
					}
				}

				proof := genProof.GetProof()
				meta := genProof.GetMetadata()
```